### DISCUSSION TITLE
Undefined reference error for example of Buildings library, link missing to function in C-Sources

### DISCUSSION THREAD

**Discussion by julicot9** [OpenModelica/OpenModelica#12813] [Q&A] [answered]
Hello,
I'm having trouble simulating the Boiler.Systems1 example from the Buildings library through OMEdit (same issue with other examples)
I'm using Linux and installed OpenModelica through the Arch user repository (see build commands here <https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=openmodelica>)
The Boiler example needs the ModelicaStandardTables_CombiTimeTable_init2 function from  ~/.openmodelica/libraries/Modelica 4.0.0+maint.om/Resources/C-Sources/ModelicaStandardTables.c but the link is not there and I get the following error when pressing the simulate button in OMEdit

```python
/usr/bin/cc -I. -o System1 System1.o System1_functions.o System1_records.o System1_01exo.o System1_02nls.o System1_03lsy.o System1_04set.o System1_05evt.o System1_06inz.o System1_07dly.o System1_08bnd.o System1_09alg.o System1_10asr.o System1_11mix.o System1_12jac.o System1_13opt.o System1_14lnz.o System1_15syn.o System1_16dae.o System1_17inl.o System1_18spd.o -L"/home/django/.openmodelica/libraries/Buildings 11.0.0/Examples/Tutorial/Boiler"  "-L/usr/bin/../lib//omc" -Wl,-rpath="/usr/bin/../lib//omc" "-L/usr/bin/../lib/" -Wl,-rpath="/usr/bin/../lib/" "-L/home/django/.openmodelica/binaries/Modelica" -Wl,-rpath="/home/django/.openmodelica/binaries/Modelica" "-L/home/django/.openmodelica/libraries/Modelica 4.0.0+maint.om/Resources/Library/x86_64-linux" -Wl,-rpath="/home/django/.openmodelica/libraries/Modelica 4.0.0+maint.om/Resources/Library/x86_64-linux" "-L/home/django/.openmodelica/libraries/Modelica 4.0.0+maint.om/Resources/Library/linux64" -Wl,-rpath="/home/django/.openmodelica/libraries/Modelica 4.0.0+maint.om/Resources/Library/linux64" "-L/home/django/.openmodelica/libraries/Modelica 4.0.0+maint.om/Resources/Library" -Wl,-rpath="/home/django/.openmodelica/libraries/Modelica 4.0.0+maint.om/Resources/Library" -lModelicaStandardTables -lModelicaIO -lModelicaMatIO -lzlib  -Os -DOM_HAVE_PTHREADS -fPIC -DOM_HAVE_PTHREADS     -I"/usr/bin/../include/omc/c" -I"/usr/bin/../include/omc" -I. -DOPENMODELICA_XML_FROM_FILE_AT_RUNTIME -DOMC_MODEL_PREFIX=System1 -DOMC_NUM_MIXED_SYSTEMS=0 -DOMC_NUM_LINEAR_SYSTEMS=2 -DOMC_NUM_NONLINEAR_SYSTEMS=0 -DOMC_NDELAY_EXPRESSIONS=0 -DOMC_NVAR_STRING=0 -L"/usr/bin/../lib//omc" -L"/usr/bin/../lib" -Wl,-rpath,"/usr/bin/../lib//omc"  -Wl,-rpath,"/usr/bin/../lib"   -lSimulationRuntimeC -lOpenModelicaRuntimeC -lomcgc -lzlib -llapack -lblas -lm -ldl -lpthread -lgfortran -lstdc++ -rdynamic  
/usr/bin/ld: System1_functions.o: in function `omc_Modelica_Blocks_Types_ExternalCombiTimeTable_constructor':
System1_functions.c:(.text+0x37e): undefined reference to `ModelicaStandardTables_CombiTimeTable_init2'
/usr/bin/ld: System1_functions.o: in function `omc_Modelica_Blocks_Tables_Internal_getNextTimeEvent':
System1_functions.c:(.text+0x1fa): undefined reference to `ModelicaStandardTables_CombiTimeTable_nextTimeEvent'
/usr/bin/ld: System1_functions.o: in function `omc_Modelica_Blocks_Tables_Internal_getTimeTableTmax':
System1_functions.c:(.text+0x213): undefined reference to `ModelicaStandardTables_CombiTimeTable_maximumTime'
/usr/bin/ld: System1_functions.o: in function `omc_Modelica_Blocks_Tables_Internal_getTimeTableTmin':
System1_functions.c:(.text+0x227): undefined reference to `ModelicaStandardTables_CombiTimeTable_minimumTime'
/usr/bin/ld: System1_functions.o: in function `omc_Modelica_Blocks_Tables_Internal_getTimeTableValueNoDer':
System1_functions.c:(.text+0x23d): undefined reference to `ModelicaStandardTables_CombiTimeTable_getValue'
/usr/bin/ld: System1_functions.o: in function `omc_Modelica_Blocks_Types_ExternalCombiTimeTable_destructor':
System1_functions.c:(.text+0x41a): undefined reference to `ModelicaStandardTables_CombiTimeTable_close'
collect2: error: ld returned 1 exit status
make: *** [System1.makefile:37: omc_main_target] Error 1
Compilation process failed. Exited with code 2.
```

The issue seems to be that the C files in the ~/.openmodelica/libraries/Modelica 4.0.0+maint.om/Resources/C-Sources folder of the Modelica library are not compiled and the link to the function therefore not created
I compiled them with gcc using the instructions in the ~/openmodelica/libraries/Modelica 4.0.0+maint.om/Resources/BuildProjects/ folder but I still got the same error
Could somebody help me with this issue ?

**Comment by casella** [comment]
@AnHeuermann, @sjoelund maybe you can give some hints to @julicot9 ?

**Reply by AnHeuermann** [reply]
The `ModelicaStandardTables_CombiTimeTable_XXX` functions should be part of *libModelicaStandardTables.so* / *libModelicaStandardTables.a* and those are shipped with OpenModelica. You could try to link to one of those: OpenModelica/lib/x86_64-linux-gnu/omc/ffi/libModelicaStandardTables.so or OpenModelica/lib/x86_64-linux-gnu/omc/libModelicaStandardTables.a.

You can check if you have one of these files and if they do export the missing functions:

```bash
> objdump -TC libModelicaStandardTables.so | grep ModelicaStandardTables_CombiTimeTable_
0000000000007430 g    DF .text  0000000000000dda  Base        ModelicaStandardTables_CombiTimeTable_nextTimeEvent
00000000000022e0 g    DF .text  00000000000000bc  Base        ModelicaStandardTables_CombiTimeTable_init2
00000000000023a0 g    DF .text  00000000000006ae  Base        ModelicaStandardTables_CombiTimeTable_init3
0000000000002b90 g    DF .text  0000000000000119  Base        ModelicaStandardTables_CombiTimeTable_close
00000000000073b0 g    DF .text  0000000000000072  Base        ModelicaStandardTables_CombiTimeTable_maximumTime
0000000000006610 g    DF .text  0000000000000d33  Base        ModelicaStandardTables_CombiTimeTable_getDer2Value
0000000000002250 g    DF .text  0000000000000084  Base        ModelicaStandardTables_CombiTimeTable_init
0000000000005940 g    DF .text  0000000000000cc6  Base        ModelicaStandardTables_CombiTimeTable_getDerValue
0000000000007350 g    DF .text  0000000000000054  Base        ModelicaStandardTables_CombiTimeTable_minimumTime
0000000000008210 g    DF .text  0000000000000360  Base        ModelicaStandardTables_CombiTimeTable_read
00000000000048b0 g    DF .text  0000000000000ea5  Base        ModelicaStandardTables_CombiTimeTable_getValue

> nm libModelicaStandardTables.a | grep ModelicaStandardTables_CombiTimeTable_
0000000000000940 T ModelicaStandardTables_CombiTimeTable_close
00000000000043c0 T ModelicaStandardTables_CombiTimeTable_getDer2Value
00000000000036f0 T ModelicaStandardTables_CombiTimeTable_getDerValue
0000000000002660 T ModelicaStandardTables_CombiTimeTable_getValue
0000000000000000 T ModelicaStandardTables_CombiTimeTable_init
0000000000000090 T ModelicaStandardTables_CombiTimeTable_init2
0000000000000150 T ModelicaStandardTables_CombiTimeTable_init3
0000000000005160 T ModelicaStandardTables_CombiTimeTable_maximumTime
0000000000005100 T ModelicaStandardTables_CombiTimeTable_minimumTime
00000000000051e0 T ModelicaStandardTables_CombiTimeTable_nextTimeEvent
0000000000005fc0 T ModelicaStandardTables_CombiTimeTable_read
```

If I remember correctly these libs depend on OpenModelica, so the compiler can't compile these functions from the source files of the Modelica standard library.

**Reply by sjoelund** [reply]
If the buildprojects in the MSL have run, or there is any other libModelicaStandardTables.{so,a}, the OpenModelica-compiled version might not be used. The way we compile it on Ubuntu, we end having /usr/lib/x86_64-linux-gnu/omc/libModelicaStandardTables.a available and if that is used, it should work fine. The ffi version of libModelicaStandardTables.so should not be found when linking the executable.

> If I remember correctly these libs depend on OpenModelica, so the compiler can't compile these functions from the source files of the Modelica standard library.

This is wrong. The only special thing is the `usertab` functionality, but it does not apply here.

**Reply by julicot9** [reply]
Thank you @AnHeuermann and @sjoelund for you answers !
Indeed when looking at the compilation command in my first message, the command seems to try to get the function from /usr/lib/omc/libModelicaStandardTables.a that does exist in my system.
When I run the nm /usr/lib/omc/libModelicaStandardTables.a I get:

```bash
> nm libModelicaStandardTables.a

ModelicaStandardTables.c.o:
nm: ModelicaStandardTables.c.o: plugin needed to handle lto object
0000000000000001 C __gnu_lto_slim

ModelicaStandardTablesUsertab.c.o:
0000000000000001 C __gnu_lto_slim
```
Do you know what this mean ?

**Reply by sjoelund** [reply, marked as answer]
Did you use `-flto` to compile OpenModelica? I don't think that will help since we don't use the flag when compiling the simulation executable.

**Reply by julicot9** [reply]
No I just compiled it through the AUR which uses the following commands:
```bash
cmake -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/usr -DOM_USE_CCACHE=OFF -DOM_OMEDIT_ENABLE_QTWEBENGINE=ON -B build .
 make -C build
 ```

**Reply by julicot9** [reply]
I solved the issue !
Instead of using the Arch user repository (AUR), I installed it following the cmake guide from github (https://github.com/OpenModelica/OpenModelica/blob/master/OMCompiler/README.Linux.md)
I don't know what caused the error, the main difference I can see with the github instructions that worked compared to the ones in the AUR that didn't is that I only used cmake and the AUR uses cmake to prepare then make to build and install

**Reply by julicot9** [reply]
Thank you for your help ! 
I'll comment on the AUR page to notify the issue

**Reply by julicot9** [reply]
I explained the problem on the Arch user repository (AUR) page of openmodelica (https://aur.archlinux.org/packages/openmodelica) and the maintainer changed the PKGBUILD to fix the issue
Installing openmodelica through the AUR also completely works for me now !
